crates/krun-server/src/server.rs

```diff
@@ -1,124 +1,123 @@
+                        },
+                    };
+                    let stream = BufStream::new(stream);
+
+                    match handle_connection(stream).await {
+                        Ok((command, mut child)) => {
+                            self.child_set.spawn(async move { (command, child.wait().await) });
+                            self.set_child_processes(self.child_set.len());
                         },
                         Err(err) => {
-                            eprintln!("Failed to wait for {command:?} process to exit: {err}");
+                            eprintln!("Failed to process client request: {err:?}");
                         },
-                    },
-                    Err(err) => {
-                        error!(err:% = err; "child task failed");
-                    },
-                }
-                state_tx.send_if_modified(|state| {
-                    let child_processes = child_set.len();
-                    if state.child_processes == child_processes {
-                        return false;
                     }
-                    state.child_processes = child_processes;
-                    true
-                });
+                    self.set_connection_idle(true);
+                },
+                Some(res) = self.child_set.join_next() => self.handle_child_join(res),
+            }
+        }
+    }
+
+    fn handle_child_join(&self, res: Result<(PathBuf, ChildResult), JoinError>) {
+        match res {
+            Ok((command, res)) => match res {
+                Ok(status) => {
+                    debug!(command:?; "child process exited");
+                    if !status.success() {
+                        if let Some(code) = status.code() {
+                            eprintln!("{command:?} process exited with status code: {code}");
+                        } else {
+                            eprintln!(
+                                "{command:?} process terminated by signal: {}",
+                                status
+                                    .signal()
+                                    .expect("either one of status code or signal should be set")
+                            );
+                        }
+                    }
+                },
+                Err(err) => {
+                    eprintln!("Failed to wait for {command:?} process to exit: {err}");
+                },
+            },
+            Err(err) => {
+                error!(err:% = err; "child task failed");
             },
         }
+        self.set_child_processes(self.child_set.len());
+    }
+
+    fn set_connection_idle(&self, connection_idle: bool) {
+        self.state_tx.send_if_modified(|state| {
+            if state.connection_idle == connection_idle {
+                return false;
+            }
+            state.connection_idle = connection_idle;
+            true
+        });
+    }
+
+    fn set_child_processes(&self, child_processes: usize) {
```

Comment:
It is of course unnecessary to take `child_processes` as a parameter. But it reads better?